dev/TreeView/ViewModel.cpp

```diff
@@ -310,15 +318,85 @@ void ViewModel::NodeCollapsed(const winrt::event_token token)
 
 void ViewModel::SelectAll()
 {
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
     UpdateSelection(m_originNode.get(), TreeNodeSelectionState::Selected);
 }
 
-void ViewModel::ModifySelectByIndex(int index, TreeNodeSelectionState const& state)
+void ViewModel::SelectItem(winrt::IInspectable const& item)    
+{
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
+    auto selectedItems = GetSelectedItems();
+    if (selectedItems.Size() > 0)
+    {
+        selectedItems.Clear();
+    }
+    if (item)
+    {
+        selectedItems.Append(item);
+    }
+}
+
+void ViewModel::SelectNode(const winrt::TreeViewNode& node, bool isSelected)
+{
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
+    auto selectedNodes = GetSelectedNodes();
+    if (isSelected)
+    {
+        if (IsInSingleSelectionMode() && selectedNodes.Size() > 0)
+        {
+            selectedNodes.Clear();
+        }
+        selectedNodes.Append(node);
+    }
+    else
+    {
+        unsigned int index;
+        if (selectedNodes.IndexOf(node, index))
+        {
+            selectedNodes.RemoveAt(index);
+        }
+    }
+}
+
+void ViewModel::SelectByIndex(int index, TreeNodeSelectionState const& state)
 {
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
     auto targetNode = GetNodeAt(index);
     UpdateSelection(targetNode, state);
 }
 
+void ViewModel::BeginSelectionChanges()
+{
+    if (!IsInSingleSelectionMode())
+    {
+        m_selectionTrackingCounter++;
```

Comment:
Why are we increasing m_selectiontrackingCounter here?

Reply:
In multi selection mode, if you have this

```
- 1
  - 1.1
  - 1.2
  - 1.3
```

and you select item 1, you'll have 4 calls to `SelectByIndex`.

The first call comes from here:

![image](https://user-images.githubusercontent.com/2237018/76533059-c5c30680-6477-11ea-8285-e0081d482136.png)

While the 3 others come from here:

![image](https://user-images.githubusercontent.com/2237018/76533097-d4112280-6477-11ea-9a61-355183ebaa26.png)

so the counter helps me detect when we're done changing the selection.

Reply:
So we are basically counting how often we enter the BeginSelection change function, and in EndSelection, we only raise the event if we are the "last" EndSelection call, do I understand that correctly?

I guess there isn't a different way to detect whether we are the last EndSelection call is there?
If there's no other way to find out whether or not we are in the last EndSelection call without counting, than I guess that's the way to go here.

Reply:
Yes you're correct. That's the best way I found so far. Maybe @kaiguo would have an alternative.

Reply:
Yeah, I think this it the way to go since ViewModel changes the selection states of all the ancestors and descendants recursively, we don't have the information of how deep the call will be beforehand.